slam-libraries/viam-cartographer/src/slam_service/slam_service.h

```diff
@@ -0,0 +1,80 @@
+#ifndef SLAM_SERVICE_H_
+#define SLAM_SERVICE_H_
+
+#include <grpc/grpc.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_context.h>
+
+#include <atomic>
+#include <string>
+
+#include "../io/draw_trajectories.h"
+#include "../io/read_PCD_file.h"
+#include "../io/submap_painter.h"
+#include "cartographer/mapping/map_builder.h"
+#include "common/v1/common.grpc.pb.h"
+#include "common/v1/common.pb.h"
+#include "service/slam/v1/slam.grpc.pb.h"
+#include "service/slam/v1/slam.pb.h"
+
+using grpc::ServerContext;
+using viam::service::slam::v1::GetMapRequest;
+using viam::service::slam::v1::GetMapResponse;
+using viam::service::slam::v1::GetPositionRequest;
+using viam::service::slam::v1::GetPositionResponse;
+using viam::service::slam::v1::SLAMService;
+
+namespace viam {
+
+static const int checkForShutdownIntervalMicroseconds = 1e5;
+static std::atomic<bool> b_continue_session{true};
+
+using SensorId = cartographer::mapping::TrajectoryBuilderInterface::SensorId;
+const SensorId kRangeSensorId{SensorId::SensorType::RANGE, "range"};
+const SensorId kIMUSensorId{SensorId::SensorType::IMU, "imu"};
+
+class SLAMServiceImpl final : public SLAMService::Service {
+   public:
+    // GetPosition returns the relative position of the robot w.r.t the "origin"
+    // of the map, which is the starting point from where the map was initially
+    // created.
+    ::grpc::Status GetPosition(ServerContext *context,
+                               const GetPositionRequest *request,
+                               GetPositionResponse *response) override;
+
+    // GetMap returns either an image or a pointcloud, depending on the MIME
+    // type requested
+    ::grpc::Status GetMap(ServerContext *context, const GetMapRequest *request,
+                          GetMapResponse *response) override;
+
+    void CreateMap();
+
+    // TODO
+    // void ProcessDataOnline(ORB_SLAM3::System *SLAM);
+    // void ProcessDataOffline(ORB_SLAM3::System *SLAM);
+    // Creates a simple map containing a 2x4x8 rectangular prism with the robot
+    // in the center, for testing GetMap and GetPosition.
+    // void ProcessDataForTesting(ORB_SLAM3::System *SLAM);
```

Comment:
I would leave this out for now. We can decide if it's useful when writing GetMap and GetPosition.

Reply:
Sorry, I just meant we can leave out ProcessDataForTesting for now! But this change is fine too, we can add ProcessDataOffline/Online back in when we get to them.